modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/tests/PublisherIntegrationTestBase.cs

```diff
@@ -0,0 +1,196 @@
+﻿namespace Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Tests {
+    using Autofac;
+    using Microsoft.Azure.IIoT.Agent.Framework;
+    using Microsoft.Azure.IIoT.Diagnostics;
+    using Microsoft.Azure.IIoT.Hub;
+    using Microsoft.Azure.IIoT.Hub.Mock;
+    using Microsoft.Azure.IIoT.Hub.Models;
+    using Microsoft.Azure.IIoT.Module;
+    using Microsoft.Azure.IIoT.Module.Framework;
+    using Microsoft.Azure.IIoT.Module.Framework.Client;
+    using Microsoft.Azure.IIoT.Module.Framework.Hosting;
+    using Microsoft.Azure.IIoT.Module.Framework.Services;
+    using Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Agent;
+    using Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Controller;
+    using Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Runtime;
+    using Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Engine;
+    using Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models;
+    using Microsoft.Azure.IIoT.OpcUa.Protocol;
+    using Microsoft.Azure.IIoT.OpcUa.Protocol.Services;
+    using Microsoft.Azure.IIoT.Serializers;
+    using Microsoft.Azure.IIoT.Utils;
+    using Microsoft.Extensions.Configuration;
+    using Moq;
+    using Opc.Ua;
+    using Serilog;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;    
+    using System.Text;
+    using System.Text.Json;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using static Microsoft.Azure.IIoT.Hub.Mock.IoTHubServices;
+
+    /// <summary>
+    /// Base class for integration testing, it connects to the server, runs publisher and injects mocked IoTHub services.
+    /// </summary>
+    public class PublisherIntegrationTestBase {
+        /// <summary>
+        /// Whether the module is running.
+        /// </summary>
+        public BlockingCollection<EventMessage> Events { get; set; } = new BlockingCollection<EventMessage>();
+
+        public PublisherIntegrationTestBase() {
+            // This is a fake but correctly formatted connection string.
+            var connectionString = $"HostName=dummy.azure-devices.net; DeviceId={Utils.GetHostName()};SharedAccessKeyName=iothubowner;SharedAccessKey=aXRpc25vdGFuYWNjZXNza2V5";
+            var config = connectionString.ToIoTHubConfig();
+
+            _typedConnectionString = ConnectionString.Parse(config.IoTHubConnString);
+            _exit = new TaskCompletionSource<bool>();            
+        }
+
+        protected async Task<List<JsonDocument>> ProcessMessages(string publishedNodesFile) {
+            // publishedNodesFile points to the local server on the same machine and port currently as tests are run one at a time (not parallel) it does not need randomly generated port numbers.
+            _ = Task.Run(() => HostPublisherAsync(Mock.Of<ILogger>(), publishedNodesFile));
+
+            while (Events.Count == 0) {
+                await Task.Delay(500);
```

Comment:
So it might break after 1 event? Are we satisifed which just one event?

Reply:
No timeouts? It can go on forever?

Reply:
Actually, maybe we can parameterize it and wait for the N number of events? Thoughts @suneetnangia ?

Reply:
The loop will break when if finds the event for the node as specified in the publishednodes file or if no such node is present erroneous response is returned (bad response). In either cases it _should_ return and not go in loop indefinitely (but if there's such situation we can always add default timeout).
Absolutely, the idea to have this in basic form first before we start to have parameters as we deem right, so if we need to receive more than 1 message or define conditions for loop, we can certainly parameterise it.